Add explicit privilege prompt to improve sudo UX

grayside opened the pull request:
Fixes #134

This change identifies that we need to sudo and adjusts logging/spinning behaviors accordingly. Needs some more testing (e.g., with verbose) and a bit more smarts in sudo detection, hence the WIP flag. Also some UI issues around the asynchronous nature of the spinner.

I am pausing here for approach feedback.

Here's what it looks like when it operates as intended:

### Paused for Input

![screen shot 2018-01-29 at 1 12 27 pm](https://user-images.githubusercontent.com/283489/35534733-45a3cbac-04f6-11e8-863c-cdd3b5beadb6.png)

### Success!

![screen shot 2018-01-29 at 1 11 34 pm](https://user-images.githubusercontent.com/283489/35534732-458a000a-04f6-11e8-841d-c734cca4915f.png)

febbraro commented:
So I would simply say `Administrative privileges needed` and nothing about a password, since the `Password` prompt does that already.  Look at this example, I ran `rig dns` twice. Once to get the initial prompt because I was unprivileged...

https://www.dropbox.com/s/m8twc771il6llhq/Screenshot%202018-01-29%2013.25.10.png?dl=0

And then once again, immediately, while I still had privs....

https://www.dropbox.com/s/vtx768cuhep9xk2/Screenshot%202018-01-29%2013.25.52.png?dl=0

It gave me "please enter your password" even though it just blasted right past it.  Claiming privs needed is a good plan and it gives us the newline needed for an unobscured prompt.  Claiming they are needed and blowing past it is really just a heads up and it is not confusing.

grayside commented:
Unfortunately here are some of the other results I'm getting, with a password needed:

![screen shot 2018-01-29 at 1 42 48 pm](https://user-images.githubusercontent.com/283489/35536077-9b66f60a-04fa-11e8-8b9e-a8a202691648.png)
![screen shot 2018-01-29 at 1 42 59 pm](https://user-images.githubusercontent.com/283489/35536078-9b7d736c-04fa-11e8-8589-4a0467d8a039.png)
![screen shot 2018-01-29 at 1 43 25 pm](https://user-images.githubusercontent.com/283489/35536079-9b9287b6-04fa-11e8-91cb-6bf50cd55575.png)

grayside commented:
Simplification helped tremendously. This is ready for final review.

I would appreciate if someone took it for a spin to confirm the timing issues are not still lurking.

My test has been:
```bash
./build/darwin/rig dns
```

tekante commented:
Looks like it's working as intended. First run through didn't need password as I had recently used sudo, second time includes the password prompt.

![privilege-prompt-test-results](https://user-images.githubusercontent.com/223075/35622386-b1d8e93c-0655-11e8-8649-a1f95f4bfe1b.png)

tekante commented on util/logger.go:
```diff
@@ -125,3 +127,21 @@ func (log *RigLogger) Verbose(format string, a ...interface{}) {
+	// This newline ensures the last status before escallation is preserved
+	// on-screen.
+	fmt.Println()
+	log.Spin("Evaluating administrative action...")
```
Just noting that things go by so quickly on my machine I never see the "Evaluating administrative action..." message.

grayside replied:
I wouldn't expect it to. I need the log.Spin() to start the spinner and the log.Warning() for the real message... let me test just using the same message for both, since that would avoid coder confusion.

tekante commented on util/slices.go:
```diff
@@ -0,0 +1,13 @@
+
+// IndexOfString is a general utility function that can find the index of a value
+// present in a string slice. The second value is true if the item is found.
+func IndexOfString(slice []string, value string) (int, bool) {
```
This is probably due to general newness to go and it's structures and syntax but it took me a while to realize that this was testing for the presence of an element in a collection of elements rather than the presence of a substring within a string. Probably baggage from character arrays and the concept of a slice of a string from Perl combined with previous indexOf type functions. I don't think it has to change I just thought I'd mention it.

This is relevant because commands of the form of `util.Command("bash", "-c", fmt.Sprintf("echo 'nameserver %s' | sudo tee /etc/resolver/vm", bridgeIP)).Run()` on 178 of dns.go would fail this check. It's not really an issue for Macs because that one is guaranteed to be proceeded by a command which will already trigger the Password prompt and stop the spinner (line 175 of dns.go). I think it might affect Linux however as the configureLinuxResolver function has no such preceding command.

I've attached an example of what this looks like on Linux.

![privilege-prompt-test-results-linux](https://user-images.githubusercontent.com/223075/35624191-3f10e0a6-065c-11e8-94c2-5bf94fa62432.png)

tekante replied:
One potential avenue around this would be to import the strings package and change the conditional on line 7 to something like
`		if elem == value || strings.Contains(elem, value + " ") {`

This appears to function properly for both mac and linux though it changes the nature of this function such that you'll probably want to rename it.

grayside replied:
There's a convention with exec.Cmd that the tokens all be individual array elements. I see that subshell situations might be different. Will adjust.

grayside commented:
In addition to addressing CJ's feedback, I took the following steps:

* Refactored Executor.ToString() to Executor.String() to be more in-line with Golang built-ins.
* Added the privilege escalation prompt to more of our command execution methods.
* When resuming the spinner, it now flashes "resuming operation", this is likely to be a flicker in many cases, but if the calling code doesn't do a good job of spinner management it may be confusing to see the spinning without context. I *think* this is a good idea, but could be convinced to leave it as an empty string.

febbraro commented:
I'm getting prompt overlap on `rig stop`

Using `sudo -k` first to remove privs

febbraro commented:
Prompt overlap is only when stopping a running machine. If I stop an already stopped machine, the prompt works fine.

grayside commented:
I haven't found a way to crack the rig stop spinner yet. It appears that the `route` command does something unusual in how it interacts with newlines or the password prompt, and may require much deeper research. Is this change a sufficient improvement to move forward anyway?

Should I output an extra message on rig stop to provide instruction on dealing with the confusion?

tekante commented:
I could go with proceeding if it is just the stop portion that overlap is wonky on. If it's something with route would an initial `sudo cat /dev/null` just so that the password gets prompted for alleviate that without causing ill effects?

grayside commented:
cat /dev/null seems to have helped. Will need to research the implications of that trick.
